File: 0940-distinct-subsequences-ii/0940-distinct-subsequences-ii.java
class Solution {
    static final long MOD = 1000000007;

    public int distinctSubseqII(String s) {
        long x = 1;

        HashMap<Character, Long> m = new HashMap<>();

        for (int i=0; i<s.length(); i++){
            char c = s.charAt(i);

            long y = x;

            x = (2 * x - m.getOrDefault(c, 0L) + MOD) % MOD;

            m.put(c, y);
        }

        return (int)((x - 1 + MOD) % MOD);
    }
}

File: 0940-distinct-subsequences-ii/0940-distinct-subsequences-ii.py
class Solution:
    MOD = 1000000007
    def distinctSubseqII(self, s: str) -> int:
        x = 1

        m = {}

        for c in s:
            y = x

            x = (2*x - m.get(c, 0) + self.MOD) % self.MOD

            m[c] = y

        return (x - 1 + self.MOD) % self.MOD